packages/evals/src/scorers/llm/context-relevance/index.ts

```
@@ -0,0 +1,164 @@
+      relevanceLevel: z.enum(['high', 'medium', 'low', 'none']),
+      wasUsed: z.boolean(),
+      reasoning: z.string(),
+    }),
+  ),
+  missingContext: z.array(z.string()).optional().default([]),
+  overallAssessment: z.string(),
+});
+
+// Default penalty constants for maintainability and clarity
+const DEFAULT_PENALTIES = {
+  UNUSED_HIGH_RELEVANCE_CONTEXT: 0.1, // 10% penalty per unused high-relevance context
+  MISSING_CONTEXT_PER_ITEM: 0.15, // 15% penalty per missing context item
+  MAX_MISSING_CONTEXT_PENALTY: 0.5, // Maximum 50% penalty for missing context
+} as const;
+
+export function createContextRelevanceScorerLLM({
+  model,
+  options,
+}: {
+  model: MastraLanguageModel;
+  options: ContextRelevanceOptions;
+}) {
+  if (!options.context && !options.contextExtractor) {
+    throw new Error('Either context or contextExtractor is required for Context Relevance scoring');
+  }
+  if (options.context && options.context.length === 0) {
+    throw new Error('Context array cannot be empty if provided');
+  }
+
+  return createScorer<ScorerRunInputForAgent, ScorerRunOutputForAgent>({
+    name: 'Context Relevance (LLM)',
+    description: 'Evaluates how relevant and useful the provided context was for generating the agent response',
+    judge: {
+      model,
+      instructions: CONTEXT_RELEVANCE_INSTRUCTIONS,
+    },
+  })
+    .analyze({
+      description: 'Analyze the relevance and utility of provided context',
+      outputSchema: analyzeOutputSchema,
+      createPrompt: ({ run }) => {
+        const userQuery = getUserMessageFromRunInput(run.input) ?? '';
+        const agentResponse = getAssistantMessageFromRunOutput(run.output) ?? '';
+
+        // Get context either from options or extractor
+        const context = options.contextExtractor ? options.contextExtractor(run.input!, run.output) : options.context!;
+
+        if (context.length === 0) {
+          throw new Error('No context available for evaluation');
```

not quite sure if we want to throw an error here. maybe we can default to 0 or 1 and give a reason that no context was available.

What do you think? if it makes more sense to throw an error, I'm okay with that as well

---

ah good point, this is a better approach.
updating!